utils/merge.go

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) 2019-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package utils
+
+import (
+	"fmt"
+	"reflect"
+)
+
+// Merge will return a new struct of the same type as base and patch, with patch merged into base.
+// Specifically, patch's values will be preferred except when patch's value is `nil`.
+// Restrictions/guarantees:
+//   - base or patch will not be modified
+//   - base and patch can be pointers or values
+//   - base and patch must be the same type
+//   - base and patch must have no unexported types (at the moment)
+//   - if maps or slices are different, the entire map or slice will be replaced (at the moment)
+//
+// Usage: callers need to cast the returned interface back into the original type, eg:
+// func mergeTestStruct(base, patch *testStruct) (*testStruct, error) {
+//	   ret, err := merge(base, patch)
+//	   if err != nil {
+//         return nil, err
+//	   }
+//	   retTS := ret.(testStruct)
+//	   return &retTS, nil
+// }
+func Merge(base interface{}, patch interface{}) (interface{}, error) {
+	if reflect.TypeOf(base) != reflect.TypeOf(patch) {
+		return nil, fmt.Errorf("cannot merge different types. base type: %s, patch type: %s",
+			reflect.TypeOf(base), reflect.TypeOf(patch))
+	}
+
+	baseType := reflect.TypeOf(base)
+	if baseType.Kind() == reflect.Ptr {
+		baseType = baseType.Elem()
+	}
+
+	ret := reflect.New(baseType)
+	if err := initStruct(baseType, ret.Elem()); err != nil {
+		return nil, err
+	}
+
+	if err := mergeRec(base, patch, ret.Elem()); err != nil {
+		return nil, err
+	}
+	return ret.Elem().Interface(), nil
+}
+
+// mergeRec recursively merges into ret
+func mergeRec(base interface{}, patch interface{}, ret reflect.Value) (err error) {
+	bt := reflect.TypeOf(base)
+	bv := reflect.ValueOf(base)
+	pv := reflect.ValueOf(patch)
+	if bt.Kind() == reflect.Ptr {
+		bt = bt.Elem()
+		bv = bv.Elem()
+		// if the patch value is nil, just assign the base value and we're done.
+		if pv.IsNil() {
+			ret.Set(bv)
+			return nil
+		}
+		pv = pv.Elem()
+	}
+	for i := 0; i < bt.NumField(); i++ {
```

Comment:
`NumField` assumes we're dealing with a struct. This method might be simpler if we moved the switch on the kind below outside the loop, and recursed on non-structs as well:
```go
switch bt.Kind() {
    case reflect.Struct: {
        for i := 0; i < bt.NumField(); i++ {
            newValue, err := mergeRec(reflect.ValueOf(bt.Field(i).Interface()))
            if err != nil {
                return errors.Wrap(err, ...)
            }
            rv.Field(i).Set(newValue)
        }
    }
    case reflect.Ptr: {
        // ...
    }
    // ...
}
```

This has the nice side effect of making the high-level `Merge` function convertible to work with non-struct types as well.